### PR TITLE
[Linux, Keyboard] Fix synthesization upon different logical keys  

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR fixes an assertion error when the physical-logical map changes during key synthesization. This error will lead to irregular key sequences, and might occur during an extremely rare edge case described in the unit test in this PR.

This PR fixes one of the two symptoms of https://github.com/flutter/flutter/issues/93278.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: shell/platform/linux/fl_key_embedder_responder.cc [hunk 1]
```diff
@@ -427,48 +444,59 @@ static void synchronize_pressed_states_loop_body(gpointer key,
...
+    any_pressed_by_record =
+        any_pressed_by_record || this_key_pressed_before_event;
+
+    if (this_key_pressed_before_event && !any_pressed_by_state) {
+      const uint64_t recorded_physical_key =
+          lookup_hash_table(self->mapping_records, logical_key);
+      // Since this key has been pressed before, there must have been a recorded
+      // physical key.
+      g_return_if_fail(recorded_physical_key != 0);
+      // In rare cases #recorded_logical_key is different from #logical_key.
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply shell/platform/linux/fl_key_embedder_responder.cc:488]: I see that the linux folder uses `#foo` as `` `foo` ``.
